### ISSUE TITLE
[feature request] Implementing Block Sparse Operations

### ISSUE THREAD

**Issue by rdspring1** [pytorch/pytorch#9222]
**TL;DR:**  Implementing block-sparse operations for faster matrix-multiplication. Is this something worth adding to PyTorch?

Goals:
1. Faster matrix-multiplication by taking advantage of block-sparsity
2. Improve running time of large-scale LSTMs - Word Language or Sentiment Analysis Models
3. Based on https://blog.openai.com/block-sparse-gpu-kernels/

New Features:
1. LinearBSMM Layer – A plug & play replacement for the standard linear layer where the user specifies the block size and sparsity pattern
2. Helper functions to generate different sparsity patterns – Random, Barabasi Albert, Watts Strogatz

Examples:
**Original Linear Layer**
```
m = nn.Linear(20, 30)
input = torch.randn(128, 20)
output = m(input)
```
**Block Sparse Linear Layer**
```
D = 20
N = 30
m = linearBSMM.random(D, N, p=0.25, block_size=32)
input = torch.randn(128, D)
output = m(input)
```
OR
```
D = 20
N = 30
sparsity_pattern = linearBSMM.random(D, N, p=0.25, block_size=32)
m = nn.LinearBSMM(D, N, sparsity_pattern)
input = torch.randn(128, D)
output = m(input)
```

**Comment by bhushan23** [comment]
This will indeed help sparser nets. But we need to scope this feature.
@soumith @SsnL @apaszke what's your view?

**Comment by zou3519** [comment]
Could be interesting, we would have to think about how this fits into the rest of sparse @li-roy @weiyangfb

**Comment by rdspring1** [comment]
Here is some more information about the underlying block-sparse operations:
- None of the inputs, outputs, or gradients use the COO sparse matrix representation.
- The sparsity pattern is a binary mask (D / block-size, N / block-size), representing the active blocks in the weight matrix.
- The block-sparse weight matrix is a dense tensor (non-zero blocks, block-size, block-size).
- The forward and back-propagation (DX, DW) operations return dense tensors.

**Comment by zou3519** [comment]
To fit it into how pytorch works currently, we could represent block-sparse tensors as 4D COO tensors, where the first two dimensions are sparse and the last two are dense. Essentially the first two dimensions indicate which "block of the grid" we're in, and the last two dimensions are the dense pieces of the matrix. I'm assuming the blocks are all the same size.

But it might be nicer to have a user-facing block sparse API

**Comment by bhushan23** [comment]
What if we support sparsity internally? I will be lot of work though but will be far more efficient and save half the space.

**Comment by weiyangfb** [comment]
another work related to block sparse: https://github.com/uber/sbnet

**Comment by mhattingpete** [comment]
Also the sparse transformer needs these operations: [https://github.com/openai/sparse_attention](https://github.com/openai/sparse_attention)

**Comment by calclavia** [comment]
Would be great to have sparsity support in general. There's a lot of interesting research towards sparser neural networks that would highly benefit from actual computational speedups rather than theoretical metrics.

**Comment by ezyang** [comment]
See also https://github.com/bamos/block

**Comment by ezyang** [comment]
@bamos What's your take on this?

We're not against accepting a PR like this, but the design seems pretty complicated, so if you're interested in implementing this, please talk to us first.

**Comment by ezyang** [comment]
@ifedan says: an intern at FAIR asked about this issue, wanting to know if we were going to implement this or not. So it might be possible that they would be interested in implementing this.

**Comment by bamos** [comment]
> @bamos What's your take on this?

This would be super-cool functionality to have in PyTorch! The scale and impacts of this could go way beyond block-sparse GPU matrix multiplications for deep learning and this will make substantial impacts to the optimization community. Most optimization problems in the real-world have a block-sparse structure that the state-of-the-art solvers, e.g. SCS/OSQP, mostly do *not* take advantage of -- they just use CPU-based sparse operations and having mixed sparse-dense support could significantly improve the performance for a non-trivial class of problems.

Here are some specific cool directions we could take this that I've thought about and brainstormed some with others over the past year (\cc @zkolter @bodono @akshayka @sbarratt @SteveDiamond @bstellato)

1. Add in custom linsys solvers to SCS/OSQP to operate purely in PyTorch with sparse/dense operations, also on the GPU or CPU (or mixed between the GPU/CPU :)). A few months ago I added in support to SCS to do the linsys solves purely in Python here (https://github.com/bodono/scs-python/pull/11) and will soon be releasing a version of SCS that uses a PyTorch backend for batched dense direct linsys solves on the GPU. If we want to go deeper here this would be a good starting point to get an upper-bound of how much mixed sparse-dense operations would help by looking at some specific problems.
2. I think a similar modification can be made to OSQP as [their linsys solver code](https://github.com/oxfordcontrol/osqp/tree/master/lin_sys) is structured similar to SCS although some serious changes may be needed to allow the data to remain completely in Python/PyTorch.
3. We could modify cvxpy to output a reasonable mixed sparse-dense format instead of just a sparse format that could then be directly input into these solvers so that we wouldn't even need to do analysis for input into these solvers

All of this said, it seems extremely non-trivial to correctly implement and properly add this functionality into PyTorch. I'd be very interested in going into more details behind this at some point.

----

> See also https://github.com/bamos/block

This is a useful prototyping tool (just a few lines of code) but probably isn't very useful here. You can use it to combine mixed sparse-dense abstract linear operators a block format, but that currently just works for scipy abstract linear operators and doesn't work with PyTorch.

**Comment by npuichigo** [comment]
@ezyang Please see also another work related to block sparse: https://github.com/mozilla/LPCNet

And TVM already supports the block sparse dense operator which has comparable performance as handwritten codes: https://docs.tvm.ai/api/python/relay/nn.html?highlight=sparse#tvm.relay.op.nn.nn.sparse_dense

**Comment by npuichigo** [comment]
@ezyang @bamos Please see also another work related to block sparse: https://github.com/mozilla/LPCNet

And TVM already supports the block sparse dense operator which has comparable performance as handwritten codes: https://docs.tvm.ai/api/python/relay/nn.html?highlight=sparse#tvm.relay.op.nn.nn.sparse_dense

**Comment by ptillet** [comment]
For those of you who are interested, I have just implemented block-sparse linear modules using the Triton language and compiler in the following repo:

https://github.com/ptillet/torch-blocksparse

**Comment by Balandat** [comment]
So gpytorch's `LazyTensor` module (https://github.com/cornellius-gp/gpytorch/tree/master/gpytorch/lazy) has some basic functionality for this. It supports things like `BlockLazyTensor`, `BlockDiagLazyTensor` and exploits this structure when doing things like matmuls, solves, cholesky decompositions, log-prob computations etc. Its focus is on psd matrices though, rather than general block-sparse problems. 

I very much hope something like this can be brought into pytorch more generally as a `LinearOperator` abstraction (#28341), possibly leveraging the proposed changes to `__torch__function__` in https://github.com/Quansight-Labs/rfcs/blob/torch_function/RFC-0001-torch-function-for-methods.md

**Comment by matthijsvk** [comment]
I'd also be very interested in Pytorch support for (block) sparse operations.
In neural networks for example, OpenAI uses it for language modeling, speech, and I think also in their DOTA2 AI.

**Comment by duncster94** [comment]
FYI, OpenAI had a blog post about a month ago about this. They're actively implementing blocksparse kernels in PyTorch.

https://openai.com/blog/openai-pytorch/

**Comment by OhadRubin** [comment]
What is the status of this issue with regards to the cuSPARSE 11.4.0 cuda toolkit released in march? 
https://developer.nvidia.com/blog/accelerating-matrix-multiplication-with-block-sparse-format-and-nvidia-tensor-cores/

**Comment by IvanYashchuk** [comment]
Experimental support for the Block Compressed Row (BSR) format was recently added in https://github.com/pytorch/pytorch/pull/68707, https://github.com/pytorch/pytorch/pull/68708, https://github.com/pytorch/pytorch/pull/68709, https://github.com/pytorch/pytorch/pull/68710 to support `torch.triangular_solve`, `torch.addmv`, `torch.addmm` on CUDA and CPU.

Currently, unsafe private methods need to be used to construct a compatible BSR input. Let's take an example matrix from [`scipy.sparse.bsr_matrix` docs](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.bsr_matrix.html)
```python
from scipy.sparse import bsr_matrix
import torch
indptr = torch.tensor([0, 2, 3, 6], device='cuda')
indices = torch.tensor([0, 2, 2, 0, 1, 2], device='cuda')
data = torch.tensor([1, 2, 3, 4, 5, 6], device='cuda').repeat(4).reshape(6, 2, 2).to(torch.float)
scipy_bsr = bsr_matrix((data.cpu(), indices.cpu(), indptr.cpu()), shape=(6, 6))
print(scipy_bsr.toarray())
# [[1. 2. 0. 0. 5. 6.]
#  [3. 4. 0. 0. 1. 2.]
#  [0. 0. 0. 0. 3. 4.]
#  [0. 0. 0. 0. 5. 6.]
#  [1. 2. 5. 6. 3. 4.]
#  [3. 4. 1. 2. 5. 6.]]
block_size = 2
torch_bsr = torch._sparse_csr_tensor_unsafe(indptr , indices, data , (3 * block_size, 3 * block_size))
print(torch_bsr)
# tensor(crow_indices=tensor([0, 2, 3, 6]),
#        col_indices=tensor([0, 2, 2, 0, 1, 2]),
#        values=tensor([[[1, 2],
#                        [3, 4]],
# 
#                       [[5, 6],
#                        [1, 2]],
# 
#                       [[3, 4],
#                        [5, 6]],
# 
#                       [[1, 2],
#                        [3, 4]],
#
#                       [[5, 6],
#                        [1, 2]],
# 
#                       [[3, 4],
#                        [5, 6]]]), size=(6, 6), nnz=6, layout=torch.sparse_csr)
# matmul works
print(torch_bsr @ torch.randn(6, 6, device='cuda'))
# tensor([[ -5.7301,   7.3167,  -2.2881,  -9.1309,  -5.7473,  -7.1322],
#         [ -0.8601,  -3.7419,   6.8649,  -4.9985,   4.9096,   3.0285],
#         [ -4.4243,   6.9137,  -3.5437,  -4.0429,  -6.1794,  -5.9879],
#         [ -6.2434,  10.6544,  -6.3818,  -7.0077,  -9.6663,  -9.3455],
#         [ -3.9079,   0.9590,  10.5623,  -9.5509,  -3.9459,  -9.7190],
#         [ -3.8839,   3.3494,   3.6550, -11.5102,  -3.0980,  -5.6420]],
#        device='cuda:0')
```
